packages/hardhat-core/src/internal/vendor/await-semaphore/index.ts

```diff
@@ -64,3 +67,154 @@ export class Mutex extends Semaphore {
     super(1);
   }
 }
+
+export class MultiProcessMutex {
+  // Logic explanation: the fs.writeFile function, when used with the wx+ flag, performs an atomic operation to create a file.
+  // If multiple processes try to create the same file simultaneously, only one will succeed.
+  // This logic can be utilized to implement a mutex.
+  // For more info check the Nomic Notion page (internal link).
+  // Timeout logic explanation:
+  // When the function passed to the mutex is executed, it must not take longer than _defaultTimeoutInMs.
+  // If it exceeds this duration, the function is terminated, and the mutex is released.
+  // In case the process running the function crashes, other processes waiting for the mutex can detect this
+  // by checking the elapsed time since the mutex was acquired. This approach helps to avoid deadlock scenarios.
+  // The _safeMarginInMs is employed here to ensure that waiting processes do not prematurely terminate the current mutex holder
+  // before it reports an error for exceeding the allowed execution time.
+  private _log: debug.Debugger;
+  private _mutexFilePath: string;
+  private _timeoutInMs: number;
+  private _defaultTimeoutInMs = 20000;
+  private _safeMarginInMs = 2000;
+  private _mutexReleaseWaitingTimeInMs = 500;
+
+  constructor(mutexName: string, timeoutInMs?: number) {
+    this._log = debug("hardhat:await-semaphore:multi-process-mutex");
+
+    this._log(`Creating mutex with name '${mutexName}'`);
+
+    this._mutexFilePath = `${os.tmpdir()}/${mutexName}.txt`;
```

Comment:
Don't use `/` to concatenate files; that can fail on Windows. Always use `path.join` or `path.resolve`. We should probably have a linter rule for this.

Reply:
Done